aten/src/ATen/core/boxing/KernelFunction.h

```diff
@@ -0,0 +1,466 @@
+   * If the kernel function was created with an unboxed function,
+   * this will call an unboxing wrapper which then calls into that
+   * unboxed function.
+   *
+   * Example:
+   *
+   * > void boxed_func(OperatorKernel*, Stack* stack) {...}
+   * > KernelFunction func = KernelFunction::makeFromBoxedFunction(&boxed_func);
+   * > Tensor result = func.callBoxed(stack);
+   *
+   * Or, with an unboxed implementation:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callBoxed(stack);
+   */
+  void callBoxed(Stack* stack) const {
+    if (C10_UNLIKELY(boxed_kernel_func_ == nullptr)) {
+      if (unboxed_kernel_func_ == nullptr) {
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on an uninitizliaed KernelFunction.");
+      } else {
+        // TODO We want to introduce the invariant that all kernels must be callable in a boxed way, then this case should be impossible.
+        TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callBoxed() on a KernelFunction that can only be called with KernelFunction::callUnboxed().");
+      }
+    }
+
+    (*boxed_kernel_func_)(getFunctor_(), stack);
+  }
+
+  /**
+   * Call the function in an unboxed way.
+   * As the "Only" in the name suggests, this only works for KernelFunctions
+   * that are backed by an unboxed kernel. If the KernelFunction was created
+   * in a boxed way, this will fail (also see KernelFunction::callUnboxed()).
+   *
+   * KernelFunction::callUnboxed() is generally better, since it will allow
+   * calling KernelFunctions that are backed by either boxed or unboxed
+   * kernels, but that one will not work for all types.
+   *
+   * Example:
+   *
+   * > KernelFunction func = KernelFunction::makeFromUnboxedLambda(
+   * >      [] (Tensor a, bool b) -> Tensor {...});
+   * > Tensor result = func.callUnboxedOnly<Tensor, Tensor, bool>(tensor1, true);
+   */
+  template<class Return, class... Args>
+  Return callUnboxedOnly(Args... args) const {
+    // note: Args above is intentionally not Args&&. We don't want perfect
+    // forwarding, which would require Args to be deduced, but instead we
+    // want callers to explicitly specify the Args.
+
+    // TODO Remove this function once all kernels support a boxed variant
+
+    TORCH_INTERNAL_ASSERT(!signature_hash_.has_value() || (detail::hashFunctionSignature<Return (Args...)>() == *signature_hash_),
+      "Called KernelFunction::callUnboxedOnly with wrong argument types");
+
+    if (unboxed_kernel_func_ != nullptr) {
+      using ActualSignature = Return (OperatorKernel*, Args...);
+      ActualSignature* func = reinterpret_cast<ActualSignature*>(unboxed_kernel_func_);
+      return (*func)(getFunctor_(), std::forward<Args>(args)...);
+    }
+
+    TORCH_INTERNAL_ASSERT(false, "Tried to call KernelFunction::callUnboxedOnly() for a kernel that doesn't have an unboxed version.");
```

Comment:
is it the only difference between callUnboxedOnly and callBoxed? why do you need two functions? is it some kind of perf optimization to simplify forwarding?

Reply:
ah, you want to avoid resolving the template for boxing, makes sense. (explanation here is clearer: https://github.com/pytorch/pytorch/pull/26273#issuecomment-531974955 )